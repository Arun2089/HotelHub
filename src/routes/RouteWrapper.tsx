import { ReactElement, useContext } from "react";
import { Navigate } from "react-router-dom";
import { AuthContext } from "@context/AuthContext";

type RouteWrapperProps = {
  components: { [key: string]: ReactElement };
};

const RouteWrapper = ({ components }: RouteWrapperProps) => {
  const {
    user: { role },
  } = useContext(AuthContext);
  return role ? components?.[role] : <Navigate to="/login" />;
};
export default RouteWrapper;
