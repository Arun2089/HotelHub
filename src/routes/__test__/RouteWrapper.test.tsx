import { render, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import AuthProviderMock from "@src/test-utils/AuthContextMock";
import RouteWrapper from "../RouteWrapper";

const adminChildComponentmock = (
  <div data-testid="admin-component">Admin Child Component</div>
);
const studentChildComponentmock = (
  <div data-testid="student-component">Student Child Component</div>
);
const userContextValues = (role: string) => {
  return {
    user: {
      id: 1,
      email: "johnDoe@gmail.com",
      role: role,
      name: "John Doe",
    },
  };
};

describe("RouteWrapper", () => {
  it("snapshot test", () => {
    const { asFragment } = render(
      <AuthProviderMock value={{ isLoading: false }}>
        <RouteWrapper
          components={{
            admin: adminChildComponentmock,
            student: studentChildComponentmock,
          }}
        />
      </AuthProviderMock>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the component based on the admin's role", () => {
    const { getByTestId } = render(
      <AuthProviderMock value={{ ...userContextValues("admin") }}>
        <RouteWrapper
          components={{
            admin: adminChildComponentmock,
            student: studentChildComponentmock,
          }}
        />
      </AuthProviderMock>
    );

    expect(getByTestId("admin-component")).toBeInTheDocument();
  });

  it("renders the component based on the student's role", () => {
    const { getByTestId } = render(
      <AuthProviderMock value={{ ...userContextValues("student") }}>
        <RouteWrapper
          components={{
            admin: adminChildComponentmock,
            student: studentChildComponentmock,
          }}
        />
      </AuthProviderMock>
    );

    expect(getByTestId("student-component")).toBeInTheDocument();
  });

  it("renders the login page if the user's role is not correct", () => {
    const { getByTestId } = render(
      <AuthProviderMock value={{ ...userContextValues("invalidRole") }}>
        <RouteWrapper
          components={{
            admin: adminChildComponentmock,
            student: studentChildComponentmock,
          }}
        />
      </AuthProviderMock>
    );

    waitFor(() => expect(getByTestId("login-page")).toBeInTheDocument());
  });
});
