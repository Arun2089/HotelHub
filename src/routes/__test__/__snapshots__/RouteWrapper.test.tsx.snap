// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteWrapper snapshot test 1`] = `
<DocumentFragment>
  <div
    data-testid="admin-component"
  >
    Admin Child Component
  </div>
</DocumentFragment>
`;
